File: src/java/Standard.java


class Standard {
  private int input_nodes;
  private int hidden_nodes;
  private int output_nodes;
  private Matrix weights_ih;
  private Matrix weights_ho;
  private double learning_rate;
  private Matrix bias_h;
  private Matrix bias_o;

  public Standard(int input_nodes, int hidden_nodes, int output_nodes) {
    this.input_nodes = input_nodes;
    this.hidden_nodes = hidden_nodes;
    this.output_nodes = output_nodes;

    this.weights_ih = new Matrix(this.hidden_nodes, this.input_nodes);
    this.weights_ho = new Matrix(this.output_nodes, this.hidden_nodes);
    this.weights_ih.randomize();
    this.weights_ho.randomize();

    this.bias_h = new Matrix(this.hidden_nodes, 1);
    this.bias_o = new Matrix(this.output_nodes, 1);
    this.bias_h.randomize();
    this.bias_o.randomize();
    this.learning_rate = 0.1;
  }

  public double sigmoid(int x){
      return 1/(1+Math.exp(-x));
  }
  public double dsigmoid(int y){
      return y*(1-y);
  }

  public double getLearningRate() {
    return this.learning_rate;
  }

  public void setLearningRate(int x) {
    this.learning_rate = x;
  }

  public void setWeights(Matrix[] x) {
    this.weights_ih = x[0];
    this.weights_ho = x[1];
  }

  public void setBiases(Matrix[] x) {
    this.biases_h = x[0];
    this.biases_o = x[1];
  }

  public Matrix getWeights(int x) {
    if (x == 0) {
      return this.weights_ih;
    } else {
      return this.weights_ho;
    }
  }
  public Matrix getBias(int x) {
    if (x == 0) {
      return this.bias_h;
    } else {
      return this.bias_o;
    }
  }

  public int predict(double[] input_array) {
    Matrix inputs = Matrix.fromArray(input_array);

    //input to hidden
    Matrix hidden = Matrix.multiply(this.weights_ih, inputs);
    hidden.add(this.bias_h);

    //activation
    hidden.map(sigmoid);

    //hidden to output
    Matrix output = Matrix.multiply(this.weights_ho, hidden);
    output.add(this.bias_o);
    output.map(sigmoid);

    //converting matrix to array and return
    return output.toArray();
  }

  public void train(input_array, target_array) {
    Matrix inputs = Matrix.fromArray(input_array);

    //input to hidden
    Matrix hidden = Matrix.multiply(this.weights_ih, inputs);
    hidden.add(this.bias_h);

    //activation
    hidden.map(sigmoid);

    //hidden to output
    Matrix outputs = Matrix.multiply(this.weights_ho, hidden);
    outputs.add(this.bias_o);
    outputs.map(sigmoid);

    Matrix targets = Matrix.fromArray(target_array);
    //error calculation
    Matrix output_errors = Matrix.subtract(targets, outputs);
    Matrix gradients = Matrix.map(outputs, dsigmoid);
    gradients.multiply(output_errors);
    gradients.multiply(this.learning_rate);

    Matrix hidden_T = Matrix.transpose(hidden);
    Matrix weight_ho_deltas = Matrix.multiply(gradients, hidden_T);

    //Weights and Biases adjustments
    this.weights_ho.add(weight_ho_deltas);
    this.bias_o.add(gradients);

    //hidden layer errors
    Matrix who_t = Matrix.transpose(this.weights_ho);
    Matrix hidden_errors = Matrix.multiply(who_t, output_errors);
    Matrix hidden_gradient = Matrix.map(hidden, dsigmoid);
    hidden_gradient.multiply(hidden_errors);
    hidden_gradient.multiply(this.learning_rate);

    Matrix inputs_T = Matrix.transpose(inputs);
    Matrix weight_ih_deltas = Matrix.multiply(hidden_gradient, inputs_T);

    this.weights_ih.add(weight_ih_deltas);
    this.bias_h.add(hidden_gradient);
  }
}

File: utils/networks/Standard/npm/Standard.js
const np = require("numpy-matrix-js");

let defaultConfig = {
  learning_rate: 0.1,
  activation: "sigmoid",
};

class Standard {
  constructor(nodes, config = defaultConfig) {
    if (nodes.length == 3) {
      this.input_nodes = nodes[0];
      this.hidden_nodes = nodes[1];
      this.output_nodes = nodes[2];
    } else {
      this.input_nodes = nodes[0];
      this.hidden_nodes = 64;
      this.output_nodes = node[1];
    }

    this.weights_ih = np.random.rand(this.hidden_nodes, this.input_nodes);
    this.weights_ho = np.random.rand(this.output_nodes, this.hidden_nodes);

    this.bias_h = np.random.rand(this.hidden_nodes, 1);
    this.bias_o = np.random.rand(this.output_nodes, 1);

    this.learning_rate = config.learning_rate;

    if (config.activation.toLowerCase() == "sigmoid") {
      this.activation = {
        function: np.sigmoid,
        dfunction: np.dsigmoid,
      };
    } else if (config.activation.toLowerCase() == "tanh") {
      this.activation = {
        function: np.tanh,
        dfunction: np.dtanh,
      };
    } else if (config.activation.toLowerCase() == "relu") {
      this.activation = {
        function: np.relu,
        dfunction: np.heaviside,
      };
    } else if (config.activation.toLowerCase() == "leakyrelu") {
      this.activation = {
        function: np.leakyrelu,
        dfunction: np.dleakyrelu,
      };
    }
  }

  getState() {
    return {
      layers: {
        input_nodes: this.input_nodes,
        hidden_nodes: this.hidden_nodes,
        output_nodes: this.output_nodes,
      },
      weights: {
        weights_ih: this.weights_ih,
        weights_ho: this.weights_ho,
      },
      biases: {
        bias_ih: this.bias_ih,
        bias_ho: this.bias_ho,
      },
      learning_rate: this.learning_rate,
    };
  }

  setState(json) {
    this.weights_ih = json.weights.weights_ih;
    this.weights_ho = json.weights.weights_ho;
    this.bias_ih = json.biases.bias_ih;
    this.bias_ho = json.biases.bias_ho;
    this.learning_rate = json.learning_rate;
    this.input_nodes = json.layers.input_nodes;
    this.hidden_nodes = json.layers.hidden_nodes;
    this.output_nodes = json.layers.output_nodes;
  }

  getLearningRate() {
    return this.learning_rate;
  }

  setLearningRate(x = 0.1) {
    this.learning_rate = x;
  }

  predict(input_array) {
    let inputs = np.zeros(input_array.length, 1);
    for (let i = 0; i < input_array.length; i++) {
      inputs[i][0] = input_array[i];
    }
    //input to hidden
    let hidden = np.matmul(this.weights_ih, inputs);
    hidden = np.add(hidden, this.bias_h);

    //activation
    hidden.map(this.activation.function);

    //hidden to output
    let output = np.matmul(this.weights_ho, hidden);
    output = np.add(output, this.bias_o);
    output.map(this.activation.function);

    //converting matrix to array and return
    return output;
  }

  train(input_array, target_array) {
    let inputs = np.zeros(input_array.length, 1);
    for (let i = 0; i < input_array.length; i++) {
      inputs[i][0] = input_array[i];
    }
    //input to hidden
    let hidden = np.matmul(this.weights_ih, inputs);
    hidden = np.add(hidden, this.bias_h);

    //activation
    hidden.map(this.activation.function);

    //hidden to output
    let outputs = np.matmul(this.weights_ho, hidden);
    outputs = np.add(outputs, this.bias_o);
    outputs.map(this.activation.function);

    let targets = np.zeros(target_array.length, 1);
    for (let i = 0; i < target_array.length; i++) {
      targets[i][0] = target_array[i];
    }
    //error calculation
    let output_errors = np.subtract(targets, outputs);
    let gradients = outputs;
    gradients.map(this.activation.dfunction);
    gradients = np.matmul(gradients, output_errors);
    gradients.multiply(this.learning_rate);

    let hidden_T = np.transpose(hidden);
    let weight_ho_deltas = np.matmul(gradients, hidden_T);

    //Weights and Biases adjustments
    this.weights_ho = np.add(this.weights_ho, weight_ho_deltas);
    this.bias_o = np.add(this.bias_o, gradients);

    //hidden layer errors
    let who_t = np.transpose(this.weights_ho);
    let hidden_errors = np.matmul(who_t, output_errors);
    let hidden_gradient = hidden;
    hidden_gradient.map(this.activation.dfunction);
    hidden_gradient = np.matmul(hidden_gradient, hidden_errors);
    hidden_gradient.multiply(this.learning_rate);

    let inputs_T = np.transpose(inputs);
    let weight_ih_deltas = np.matmul(hidden_gradient, inputs_T);

    this.weights_ih = np.add(this.weights_ih, weight_ih_deltas);
    this.bias_h = np.add(this.bias_h, hidden_gradient);
  }
}

module.exports = Standard;
